### ISSUE TITLE
Refactor features_means

### ISSUE THREAD

**Issue by MichaelCurrie** [openworm/movement_cloud#38]
1. Create a `features` table containing descriptions of each feature (possibly based on https://github.com/openworm/open-worm-analysis-toolbox/blob/master/open_worm_analysis_toolbox/features/feature_metadata/features_list.csv)
2. Have features_means reference this features table as a foreign key
3. Add information about a "top 10" features specifically referencing the "default core features"
4. Replace https://github.com/openworm/movement_cloud/blob/dev/webworm/defaultCoreFeatures.json with a view on the `features` table.

**Comment by MichaelCurrie** [comment]
`create table features (name VARCHAR(50), desc VARCHAR(200), is_core_feature BIT);`

**Comment by MichaelCurrie** [comment]
@cheelee as an illustration, here's python code to get the core features from the new database table:

```sql
import mysql.connector as sql
import pandas as pd

password_str = passwords.passwords_dict['mcurrie_mysql']

db_connection = sql.connect(host='localhost', database='mrc_db4', user='mcurrie_mysql', password=password_str)
db_cursor = db_connection.cursor()
db_cursor.execute("SELECT name FROM features WHERE is_core_feature = 1 OR `name` = 'nosuchname';")
table_rows = db_cursor.fetchall()
df = pd.DataFrame(table_rows)
print(df)
```

**Comment by cheelee** [comment]
Ok at this stage, this will require the attention of the three of us - 
1. @ver228 - would you please merge @MichaelCurrie 's new "features" table (renamed to "core_features"?) to the cumulative test database we have (merge_worm_db_111017_1622.sql)?
2. I can independently work on a version of mrc_db4 on a local branch with the necessary Django code to process the core features table.

**Comment by ver228** [comment]
The merge_worm_db_111017_1622 is a dump of my local database. We can delete it. 
What do you think is the best approach? Should we create a development DB and make any update directly to AWS or should I do modifications locally and merge them later? At this point any change should be data updates.  Hopefully no more changes to the schema.

**Comment by cheelee** [comment]
Ok I think I got confused previously. So the current database mrc_db4 is the fully merged database with all the features and data we need?

If that is the case, I will need to redo what I did using a mysqldump of mrc_db4 on my local machine. An .sql file was what I was expecting earlier, instead of the actual database. Sorry about that!

**Comment by cheelee** [comment]
@ver228 On the issue of schema changes, we may still eventually need to consider changes if we need to deal with performance issues. I'm currently not too concerned about performance just yet.

**Comment by cheelee** [comment]
Ok I think I'm starting to get a better understanding about all of this now - the "features" table is primarily a way to store information about a feature's description, along with whether or not that feature is a core feature.

@MichaelCurrie Does this mean that the features table must be pre-generated using the fields of the features_means table, and then populated (description + whether or not it is core) using an external data source (e.g. the JSON file?) This also means we do not have to determine if a field is missing from the database, do we? I think this workflow should be fine as long as information in the features table is optional (i.e. will not break anything if missing) - the only thing is that the interface may either have to trust that the information in the database is consistent (e.g. rows in the features table must be found in the fields of the features_means table;) or conduct its own check at the interface code.

**Comment by cheelee** [comment]
@MichaelCurrie @ver228 I've modified the Django tool to pull core features information directly from the database now. This is now on the live development server.

**Comment by MichaelCurrie** [comment]
> Does this mean that the features table must be pre-generated using the fields of the features_means table, and then populated (description + whether or not it is core) using an external data source (e.g. the JSON file?) 

Yes, that's exactly what I did.  But we only need to do this once.

> This also means we do not have to determine if a field is missing from the database, do we?

Yes to be 100% we should have an audit step that throws an error if someone adds a feature mean that does not have an entry in the `features` table.  But as you say this kind of checking is probably too fancy for us for now.

> the only thing is that the interface may either have to trust that the information in the database is consistent (e.g. rows in the features table must be found in the fields of the features_means table;) or conduct its own check at the interface code.

Yep.  At some point we'll want to start unit testing the database in various ways.  But for now, I think we are okay.

**Comment by cheelee** [comment]
Should we discuss this again at some point?
